third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1221,4 +1515,257 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+      }
+
+      // iteration domain
+      int axisID = -1;
+      if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+        outer_reduction_tv->split(
+            axisID, rparams.vectorization_factor_tmp_gmem_read);
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::Unroll);
+      }
+
+      if (rparams.tidx_for_outer_reduction) {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDy));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDy);
+      } else {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDx));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDx);
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    } else {
+      // reduction domain
+      outer_reduction_tv->split(
+          0, NamedScalar::getParallelDim(ParallelType::TIDy));
+      outer_reduction_tv->axis(1)->parallelize(ParallelType::TIDy);
+      // iteration domain
+      int axisID = -1;
+      if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+        outer_reduction_tv->split(
+            axisID, rparams.vectorization_factor_tmp_gmem_read);
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::Unroll);
+      }
+      if (rparams.lparams.bdimx() > 1) {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDx));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDx);
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    }
+
+    auto outer_reference_tv =
+        reduction_scheduler_utils::sortAndRFactor(outer_reduction_tv);
+    outer_reference_tvs.emplace_back(outer_reference_tv);
+  }
+
+  for (auto output : dummy_outputs) {
+    fusion->addOutput(output);
+  }
+
+  const bool vectorize =
+      rparams.vectorize_inner_reduction || rparams.vectorize_iter_dom;
+  const bool is_outer_grid_persistence = rparams.persistent_kernel &&
+      rparams.cross_grid_inner_reduction && !rparams.fastest_dim;
+
+  // Propagate inner reduction
+  reduction_scheduler_utils::propagateTransformation(
+      inner_reference_tv, boundaryNodesSet);
+  reduction_scheduler_utils::propagateRFactor(
+      inner_reference_tv, inner_reduction_tv, inner_reduction_tvs);
+  reduction_scheduler_utils::propagateParallelization(
+      fusion,
+      inner_reduction_tv,
+      inner_reference_tv,
+      unroll,
+      vectorize,
+      is_outer_grid_persistence,
+      inner_reduction_tvs,
+      cached_inputs,
+      cached_outputs);
+
+  // Propagate outer reduction
+  for (long unsigned int i = 0; i < outer_reference_tvs.size(); i++) {
```

Comment:
This seems strange to me. Why is the propagation done from all of the outer reference tvs with the same boundary node set?

Reply:
added comments.
```
  // Propagate outer reduction. Each outer reduciton is conected with its
  // cached_gmem and output, since we added all the cached_gmem to the
  // boundaryNodesSet, the transformation from one outer reduction can't
  // propagate to other outer reductions due to the cutoff at boundaryNodesSet.
  // Thus, we need a loop to initiate the propagation from each outer reduction.
```

Reply:
This assumes that each of the outer reductions are connected only with the temp buffer, right? If there's an expression using the outputs of multiple outer reductions, the transforms of one outer reduction group could be propagated through it.

I think this shouldn't happen as `hasSharedConsumer` should exclude fusions with such a pattern, although I thought the intent of the `hasSharedConsumer` was just to detect shared consumers between the inner and outer reductions, not necessarily within multiple outer reductions.

Please confirm if my understanding is correct.